internal/ssh/server.go

```diff
@@ -0,0 +1,220 @@
+	"regexp"
+	"runtime"
+	"strconv"
+	"strings"
+	"sync"
+	"syscall"
+	"unsafe"
+
+	"github.com/gliderlabs/ssh"
+	gossh "golang.org/x/crypto/ssh"
+
+	"github.com/creack/pty"
+)
+
+func newServer(ca string) *ssh.Server {
+	ssh.Handle(func(s ssh.Session) {
+		user, err := user.Lookup(s.User())
+		if err != nil {
+			log.Printf("could not find user: %s", err)
+			return
+		}
+
+		shell, err := getShell(user)
+		if err != nil {
+			log.Printf("could not get user shell: %s", err)
+			return
+		}
+
+		log.Printf("new ssh session for user %s\n", s.User())
+
+		uid, _ := strconv.ParseUint(user.Uid, 10, 32)
+		gid, _ := strconv.ParseUint(user.Gid, 10, 32)
+
+		var cmd exec.Cmd
+		if len(s.Command()) > 0 {
+			cmd.Path = shell
+			cmd.Args = []string{shell, "-c", s.RawCommand()}
+		} else {
+			cmd.Path = shell
+			cmd.Args = []string{"-" + filepath.Base(shell)}
+		}
+
+		cmd.Env = []string{
+			"LANG=en_US.UTF-8",
+			"HOME=" + user.HomeDir,
+			"USER=" + user.Username,
+			"SHELL=" + shell,
+		}
+
+		cmd.Dir = user.HomeDir
+
+		ptyReq, winCh, isPty := s.Pty()
+		if isPty {
+			cmd.Env = append(cmd.Env, fmt.Sprintf("TERM=%s", ptyReq.Term))
+
+			f, err := pty.StartWithAttrs(&cmd, &pty.Winsize{}, &syscall.SysProcAttr{
+				Credential: &syscall.Credential{
+					Uid:         uint32(uid),
+					Gid:         uint32(gid),
+					NoSetGroups: true,
+				},
+				Setsid:  true,
+				Setctty: true,
+			})
+			if err != nil {
+				log.Println(err)
+				return
+			}
+			go func() {
+				for win := range winCh {
+					setWinsize(f, win.Width, win.Height)
+				}
+			}()
+			go func() {
+				io.Copy(f, s)
+			}()
+			io.Copy(s, f)
+			cmd.Wait()
+		} else {
+			cmd.SysProcAttr = &syscall.SysProcAttr{
+				Credential: &syscall.Credential{
+					Uid:         uint32(uid),
+					Gid:         uint32(gid),
+					NoSetGroups: true,
+				},
+				Setsid: true,
+			}
+
+			stdout, err := cmd.StdoutPipe()
+			if err != nil {
+				log.Printf("failed to set stdout: %v\n", err)
+				return
+			}
+			stderr, err := cmd.StderrPipe()
+			if err != nil {
+				log.Printf("failed to set stderr: %v\n", err)
+				return
+			}
+			stdin, err := cmd.StdinPipe()
+			if err != nil {
+				log.Printf("failed to set stdin: %v\n", err)
+				return
+			}
+
+			wg := &sync.WaitGroup{}
+			wg.Add(2)
+			if err = cmd.Start(); err != nil {
+				log.Printf("failed to start command %v\n", err)
+				return
+			}
+			go func() {
+				defer stdin.Close()
+				if _, err := io.Copy(stdin, s); err != nil {
+					log.Printf("failed to write to session %s\n", err)
+				}
+			}()
+			go func() {
+				defer wg.Done()
+				if _, err := io.Copy(s, stdout); err != nil {
+					log.Printf("failed to write to stdout %s\n", err)
+				}
+			}()
+			go func() {
+				defer wg.Done()
+				if _, err := io.Copy(s.Stderr(), stderr); err != nil {
+					log.Printf("failed to write from stderr%s\n", err)
+				}
+			}()
+
+			wg.Wait()
+			cmd.Wait()
+
+		}
+	})
+
+	return &ssh.Server{
+		Version: "Mysocketctl-ssh-server",
```

Comment:
I think this needs to start with SSH-2.0-
Maybe use: SSH-2.0-MySocketctl-ssh-server

Reply:
This ssh module will prepend this automatically.